thegraph/src/types/primitives/decimal.rs

```diff
@@ -0,0 +1,253 @@
+use std::{fmt, iter, str};
+
+use alloy_primitives::U256;
+
+const ONE_18: u128 = 1_000_000_000_000_000_000;
+
+/// Represents a positive decimal value with 18 fractional digits precision. Using U256 as storage.
+#[derive(Copy, Clone, Default, Hash, PartialEq, Eq, PartialOrd, Ord)]
+pub struct UDecimal18(U256);
+
+impl From<U256> for UDecimal18 {
+    fn from(value: U256) -> Self {
+        Self(U256::from(value) * U256::from(ONE_18))
+    }
+}
+
+impl From<u128> for UDecimal18 {
+    fn from(value: u128) -> Self {
+        Self::from(U256::from(value))
+    }
+}
+
+impl TryFrom<f64> for UDecimal18 {
+    type Error = <U256 as TryFrom<f64>>::Error;
+    fn try_from(value: f64) -> Result<Self, Self::Error> {
+        U256::try_from(value * 1e18).map(Self)
+    }
+}
+
+impl From<UDecimal18> for f64 {
+    fn from(value: UDecimal18) -> Self {
+        f64::from(value.0) * 1e-18
+    }
+}
+
+#[derive(Copy, Clone, Debug, Eq, PartialEq)]
+pub struct InvalidDecimalString;
```

Comment:
I just noticed that this type is a struct representing an error but it does not implement the `std::error::Error` trait 👀 

It's not blocking for this PR, but it's definitely something to fix in the near future